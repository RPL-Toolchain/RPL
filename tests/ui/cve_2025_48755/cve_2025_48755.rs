//@revisions: normal
//@[normal]compile-flags: -Z inline-mir=false
use std::{
    alloc::{alloc_zeroed, dealloc, Layout},
    mem::size_of,
    ops::{Index, IndexMut},
    slice::{from_raw_parts, from_raw_parts_mut},
};

pub struct AlignedMemory<const ALIGN: usize> {
    p: *mut u64,
    sz_u64: usize,
    layout: Layout,
}

impl<const ALIGN: usize> AlignedMemory<{ ALIGN }> {
    pub fn new(sz_u64: usize) -> Self {
        let sz_bytes = sz_u64 * size_of::<u64>();
        let layout = Layout::from_size_align(sz_bytes, ALIGN).unwrap();

        let ptr;
        unsafe {
            ptr = alloc_zeroed(layout);
            //~^ERROR: public function `new` allocates a pointer that may be zero-sized, which is an undefined behavior
        }

        Self {
            p: ptr as *mut u64,
            sz_u64,
            layout,
        }
    }
}
