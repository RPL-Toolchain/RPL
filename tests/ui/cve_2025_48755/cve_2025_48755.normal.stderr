error: public function `new` allocates a pointer that may be zero-sized, which is an undefined behavior
  --> tests/ui/cve_2025_48755/cve_2025_48755.rs:23:19
   |
LL |     pub fn new(sz_u64: usize) -> Self {
   |                ------ size passed in here
...
LL |             ptr = alloc_zeroed(layout);
   |                   ^^^^^^^^^^^^^^^^^^^^ pointer created here
   |
   = note: See https://doc.rust-lang.org/std/alloc/fn.alloc.html and https://doc.rust-lang.org/std/alloc/trait.GlobalAlloc.html#tymethod.alloc
   = note: `#[deny(rpl::alloc_maybe_zero)]` on by default

error: aborting due to 1 previous error

